src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
-          components::ContactSensorData::typeId))
-      {
-        this->contactSensorEntity = colEntity;
-
-        this->modelName = this->model.Name(_ecm);
-
-        igndbg << "Contactensor detected within model "
-            << this->modelName << std::endl;
-      }
+      this->dataPtr->forceRadius = _sdf->Get<double>("force_radius");
     }
+  }
 
-    // Link should have 1 Contact sensor and 1 Depth Camera sensor
-    auto sensorsInsideLink =
-        _ecm.ChildrenByComponents(allLinks[0], components::Sensor());
-
-    bool depthCameraInLink = false;
-    bool contactSensorInLink = false;
-    sdf::ElementPtr depthCameraSdf = nullptr;
-    for (const Entity &sensor : sensorsInsideLink)
+  // <force_length>
+  if (!_sdf->HasElement("force_length"))
+  {
+    ignlog << "Missing parameter <force_length>, "
+      << "setting to " << this->dataPtr->forceLength << std::endl;
+  }
+  else
+  {
+    if (_sdf->Get<double>("force_length") < 0)
     {
-      if (_ecm.EntityHasComponentType(sensor,
-                                      components::DepthCamera::typeId))
-      {
-        depthCameraInLink = true;
-        depthCameraSdf =
-          _ecm.Component<components::DepthCamera>(sensor)->Data().Element();
-      }
-
-      if (_ecm.EntityHasComponentType(sensor,
-                                      components::ContactSensor::typeId))
-      {
-        contactSensorInLink = true;
-      }
+      ignwarn << "Parameter <force_length> must be positive, "
+        << "setting to " << this->dataPtr->forceLength << std::endl;
     }
-
-    if (!depthCameraInLink || !contactSensorInLink ||
-      sensorsInsideLink.size() != 2)
+    else
     {
-      ignerr << "Link must have 1 Depth Camera sensor and "
-        << "1 Contact sensor" << std::endl;
-      return;
+      this->dataPtr->forceLength = _sdf->Get<double>("force_length");
     }
+  }
+
+  // Get the size of the sensor from the SDF
+  // If there's no <collision> specified inside <link>, a default one
+  // is set
+  this->dataPtr->sensorSize =
+    _sdf->GetParent()->GetElement("link")->GetElement("collision")->
+    GetElement("geometry")->GetElement("box")->
+    Get<ignition::math::Vector3d>("size");
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::PreUpdate(const UpdateInfo &_info,
+  EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("TouchPluginPrivate::PreUpdate");
+
+  // Nothing left to do if paused
+  if (_info.paused)
+    return;
 
-    // Store depth camera update rate and offset from model
-    auto offset = depthCameraSdf->Get<ignition::math::Pose3d>("pose");
-    this->cameraUpdateRate = 1;
-    if (depthCameraSdf->HasElement("update_rate"))
+  if (!this->dataPtr->initialized)
+  {
+    // We call Load here instead of Configure because we can't be guaranteed
+    // that all entities have been created when Configure is called
+    this->dataPtr->Load(_ecm);
+    this->dataPtr->initialized = true;
```

Comment:
In that case, if it fails to initialize it will continuously try to `Load()` and printing an error/warning many times per second.
On the other hand, if `initialize` is set like it currently is, it will print the error/warning once and will continue running, probably failing or doing weird things.

Another option is using another flag, e.g. `triedToLoad` to make sure we only `Load()` once. That way, if the plugin fails to load, the error/warning will only be printed once, and after that the plugin will do nothing.

Reply:
It's typical practice to have a class Boolean variable e.g. `initErrorPrinted`, so that it only prints once. In the printout, you can add in parentheses `(only printed once)`. You don't need to limit how many times `Load()` is called, because something could have changed dynamically that make later calls to `Load()` succeed, e.g. someone injects into the SDF or changes the components in the ECM.

Reply:
e216bb2